batch fetching status lists (#141)
oauth-wg/draft-ietf-oauth-status-list: draft-ietf-oauth-status-list.md

* `status_list`: REQUIRED. JSON Object that contains a Status List. The object contains exactly two claims:
   * `bits`: REQUIRED. JSON Integer specifying the number of bits per Referenced Token in the Status List (`lst`). The allowed values for `bits` are 1,2,4 and 8.
   * `lst`: REQUIRED. JSON String that contains the status values for all the Referenced Tokens it conveys statuses for. The value MUST be the base64url-encoded (as defined in Section 2 of {{RFC7515}}) Status List as specified in [](#status-list).

>    * `aggregation_uri`: OPTIONAL. An URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.

Reviewer: ```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```

>    * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve a the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.

Author: ```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

* `status_list`: REQUIRED. JSON Object that contains a Status List. It MUST contain at least the following claims:
   * `bits`: REQUIRED. JSON Integer specifying the number of bits per Referenced Token in the Status List (`lst`). The allowed values for `bits` are 1,2,4 and 8.
   * `lst`: REQUIRED. JSON String that contains the status values for all the Referenced Tokens it conveys statuses for. The value MUST be the base64url-encoded (as defined in Section 2 of {{RFC7515}}) Status List as specified in [](#status-list).

>    * `aggregation_uri`: OPTIONAL. JSON String that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.

Reviewer 2: ```suggestion
   * `aggregation_uri`: OPTIONAL. JSON String that contains a URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

* The `StatusList` structure is a map (Major Type 5) and defines the following entries:
  * `bits`: REQUIRED. Unsigned int (Major Type 0) that contains the number of bits per Referenced Token in the Status List. The allowed values for `bits` are 1, 2, 4 and 8.
  * `lst`: REQUIRED. Byte string (Major Type 2) that contains the Status List as specified in [](#status-list-json).

>   * `aggregation_uri`: OPTIONAL. Byte string (Major Type2) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.

Reviewer: Why are we repeating the definitions for CBOR Status List? In the CWT Status List Token we just reference to the JWT.Should we do the same here?

Author: Hmm, we need to define the major type here, but yeah we can probably make this a bit cleaner - let's briefly discuss tomorrow

Reviewer 2: I think we should use major type 3 - text string.

>   * `aggregation_uri`: OPTIONAL. Byte string (Major Type2) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.

Reviewer: ```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type3) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.
```

>   * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this credential type. See section [](#batch-fetching) for further detail.

Author: ```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

* The `StatusList` structure is a map (Major Type 5) and defines the following entries:
  * `bits`: REQUIRED. Unsigned int (Major Type 0) that contains the number of bits per Referenced Token in the Status List. The allowed values for `bits` are 1, 2, 4 and 8.
  * `lst`: REQUIRED. Byte string (Major Type 2) that contains the Status List as specified in [](#status-list-json).

>   * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains an URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.

Reviewer 2: ```suggestion
  * `aggregation_uri`: OPTIONAL. Text string (Major Type 3) that contains a URI to retrieve the Status List Aggregation for this type of Referenced Token. See section [](#batch-fetching) for further detail.
```

TBD

# Status List Aggregation {#batch-fetching}

> Status List Aggregation is an optional mechanism to retrieve a list of URIs to all Status List Tokens, allowing a Relying Party to fetch all relevant Status Lists for a specific type of Referenced Token or issuer. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status.

Reviewer 2: ```suggestion
Status List Aggregation is an optional mechanism to retrieve a list of URIs to all Status List Tokens, allowing a Relying Party to fetch all relevant Status Lists for a specific type of Referenced Token or issuer. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status of a reference token.
```

TBD

> # Batch Fetching {#batch-fetching}

Author: ```suggestion
# Status List Aggregation {#batch-fetching}
```

An issuer MAY support any of these mechanisms:

- Issuer metadata: The issuer of the Referenced Token publishes an URI which links to Status List Aggregation, e.g. in publicly available metadata of an issuance protocol
- Status List Parameter: The issuer of the Referenced Token includes an additional claim to the Status List (Token) that links to the Status List Aggregation.

## Issuer Metadata

The issuer MAY link to the Status List Aggregation URI in metadata that can be provided by different means like .well-known metadata as is used commonly in OAuth and OpenID, or via a VICAL extension for ISO mDoc / mDL.

The concrete specification on how this is implemented depends on the specific ecosystem and is out of scope of this specification.

## Status List Parameter

The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party may use this link to retrieve an up-to-date list of relevant Status Lists.

## Status List Aggregation in JSON Format

This section defines the structure for a JSON-encoded Status List Aggregation:

> * `status_lists`: REQUIRED. The `status_lists` claim MUST be an array of strings, each containing an URI to a Status List (Token).

Reviewer: ```suggestion
* `status_lists`: REQUIRED. JSON array of strings that contains URIs linking to Status List (Tokens).
```

TBD

# Batch Fetching {#batch-fetching}

> To allow a Relying Party to fetch all status lists for a specific type of credential or a specific issuer, an optional mechanism is provided to retrieve an Object containing a list of Status Lists called Status List Aggregation. This mechanism is intended to optimize fetching and caching mechanisms and allow offline validation of the status.

Reviewer: ```suggestion
To allow a Relying Party to fetch all Status Lists for a specific Referenced Token or issuer, an optional mechanism is provided to retrieve a list of URIs to all Status List Tokens called Status List Aggregation. This mechanism is intended to support fetching and caching mechanisms and allow offline validation of the status.
```

> There are two options that a Relying Party can use to discover the relevant Status List Aggregation URI.

Reviewer: ```suggestion
There are two options for a Relying Party to retrieve the Status List Aggregation.
```

An issuer MAY support any of these mechanisms:

> - Issuer metadata: The issuer publishes an URI which links to Status List Aggregation

Reviewer: ```suggestion
- Issuer metadata: The issuer of the Referenced Token publishes an URI which links to Status List Aggregation, e.g. in publicly available metadata of an issuance protocol
```

> - Status List Parameter: The issuer can choose to add an additional claim to the Status List (Token) that links to the Status List Aggregation.

Reviewer: ```suggestion
- Status List Parameter: The issuer of the Referenced Token includes an additional claim to the Status List (Token) that links to the Status List Aggregation.
```

> 

Reviewer: maybe turn these chapters around? Now that we mentioned the two mechanisms, giving the details on those fits better

## Status List Aggregation in JSON Format

This section defines the structure for a JSON-encoded Status List Aggregation:

> * `status_lists`: REQUIRED. Contains an array of strings, each containing an URI to a Status List (Token)

Reviewer: ```suggestion
* `status_lists`: REQUIRED. The `status_lists` claim MUST be an array of strings, each containing an URI to a Status List (Token).
```

> The Status List Aggregation URI provides a list of Status List URIs. This aggregation in JSON and the media type return SHOULD be `application/json`.

Reviewer: maybe explain the fetching logic and move the the media type to the Section 8?

The following is a non-normative example for media type `application/json`:

~~~ json

{
   "status_lists" : [
      "https://example.com/statuslists/1",
      "https://example.com/statuslists/2",
      "https://example.com/statuslists/3"
   ]
}
~~~

## Issuer Metadata

The issuer MAY link to the Status List Aggregation URI in metadata that can be provided by different means like .well-known metadata as is used in OAuth and OpenID, or via a VICAL extension for ISO mDoc / mDL.

> OpenID4VCI Example

> VICAL example

## Status List Parameter

> The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party can choose to follow this link and retrieve an up-to-date list of the relevant Status Lists.

Reviewer: ```suggestion
The URI to the Status List Aggregation MAY be provided as the optional parameter `aggregation_uri` in the Status List itself as explained in[](#status-list-cbor) and [](#status-list-json) respectively. A Relying Party may use this link to retrieve an up-to-date list of relevant Status Lists.
```
